{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "library(RPostgreSQL)\n",
    "library(dplyr)\n",
    "library(GenomicRanges)\n",
    "library(doParallel)\n",
    "library(dbplyr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "source(\"../BDDS/footprints/testdb/src/dbFunctions.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# connect to databases\n",
    "\n",
    "# load chipseq traditionally -- will just put this in memory eventually\n",
    "if(!exists(\"db.chipseq\"))\n",
    "    dbname=\"chipseq\"\n",
    "    port=\"5432\"\n",
    "    driver=dbDriver(\"PostgreSQL\")\n",
    "    user=\"trena\"\n",
    "    password=\"trena\"\n",
    "    host=\"whovian\"\n",
    "    db.chipseq <- dbConnect(drv=driver, user=user, password=password, dbname=dbname, host=host, port=port)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we've connected to our database of ChipSeq data; it's our \"gold standard\" for comparison and it doesn't contain that many different different TFs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create map: M[TF] = c(motif1,motif2,...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Auto-disconnecting PostgreSQLConnection\n"
     ]
    }
   ],
   "source": [
    "# chipseq hits table can fit in memory so load it up\n",
    "chipseq.hits <- dbGetQuery(db.chipseq, \"select * from hits\")\n",
    "chipseq.hits <- as.tbl(chipseq.hits)\n",
    "\n",
    "# get chipseq regions data and change chrom notations from eg chr10 to just 10 as in fimo\n",
    "chipseq.regions <- dbGetQuery(db.chipseq, \"select * from regions\")\n",
    "chipseq.regions <- as.tbl(chipseq.regions)\n",
    "chr.list <- chipseq.regions$chrom\n",
    "cutoff <- nchar(\"chr\")+1\n",
    "no.chr.list <- substring(chr.list,cutoff)\n",
    "chipseq.regions$chrom <- no.chr.list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# see which of the TFs in the chipseq data are mapped to any motifs\n",
    "TF.motif.pairs <- read.csv(\"text_data_files/motif_to_tf_mappings_with_tfclass_include_multiple.csv\",\n",
    "                            stringsAsFactors=FALSE)\n",
    "\n",
    "unique.cs.tfs <- unique(chipseq.hits$name)\n",
    "cs.tf.nomatches <- unique.cs.tfs[!(unique(unique.cs.tfs) %in% unique(TF.motif.pairs$tfs))]\n",
    "cs.tf.matches <- unique.cs.tfs[unique(unique.cs.tfs) %in% unique(TF.motif.pairs$tfs)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "77"
      ],
      "text/latex": [
       "77"
      ],
      "text/markdown": [
       "77"
      ],
      "text/plain": [
       "[1] 77"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "17"
      ],
      "text/latex": [
       "17"
      ],
      "text/markdown": [
       "17"
      ],
      "text/plain": [
       "[1] 17"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "60"
      ],
      "text/latex": [
       "60"
      ],
      "text/markdown": [
       "60"
      ],
      "text/plain": [
       "[1] 60"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique.cs.tfs)\n",
    "length(cs.tf.nomatches)\n",
    "length(cs.tf.matches)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Up to this point we've done this:\n",
    "\n",
    "1. Read all our hits from ChipSeq \n",
    "2. Read in all our Motif-TF mappings\n",
    "3. Pulled out the unique TFs from ChipSeq (there are 77) and found those that are in our Motif-TF mapping (there are 60)\n",
    "\n",
    "Now we want to use our motif-TF mapping to make a list of all TFs and their motifs AND to make a list of all motifs and their TFs. This will necessarily be 2 lists of lists; as Rory notes, a dictionary would be better, but this is R, not Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# create a list of lists (poor sub for dict) for TFs -> motifs and motifs -> TFs\n",
    "# only use jaspar motifs\n",
    "\n",
    "TF.motif.pairs <- as.tbl(read.csv(\"text_data_files/motif_to_tf_mappings_with_tfclass_include_multiple.csv\",\n",
    "                            stringsAsFactors=FALSE))\n",
    "\n",
    "TF.motif.pairs.jaspar <- subset(TF.motif.pairs, grepl(\"^MA[0-9]\", motif) )\n",
    "\n",
    "TFs.to.motifs <- list()\n",
    "for (TF in cs.tf.matches) {\n",
    "    this.TF.df <- subset(TF.motif.pairs.jaspar, tfs %in% TF)\n",
    "    TFs.to.motifs[TF] <- list(this.TF.df$motif)\n",
    "}\n",
    "TFs.to.motifs <- TFs.to.motifs[lapply(TFs.to.motifs,length) >= 1]\n",
    "\n",
    "motifs.to.TFs <- list()\n",
    "for (mtf in unique(TF.motif.pairs.jaspar$motif)) {\n",
    "    this.motif.df <- subset(TF.motif.pairs.jaspar, motif %in% mtf)\n",
    "    motifs.to.TFs[mtf] <- list(this.motif.df$tfs)\n",
    "}\n",
    "motifs.to.TFs <- motifs.to.TFs[lapply(motifs.to.TFs,length) >= 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "55"
      ],
      "text/latex": [
       "55"
      ],
      "text/markdown": [
       "55"
      ],
      "text/plain": [
       "[1] 55"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "514"
      ],
      "text/latex": [
       "514"
      ],
      "text/markdown": [
       "514"
      ],
      "text/plain": [
       "[1] 514"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(TFs.to.motifs)\n",
    "length(motifs.to.TFs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What this tells us is that when we look at only JASPAR motifs:\n",
    "\n",
    "* There are only 55 TFs in our ChipSeq set that are JASPAR motifs\n",
    "* There are 514 unique JASPAR motifs in our TF-motif mapping\n",
    "\n",
    "We'll save these to a couple maps that we can load/use later. We also don't actually know how many motifs are mapped to by our TFs; we only know the total number of JASPAR motifs and the total number of TFs that have JASPAR motifs. \n",
    "\n",
    "We'll quickly run through our TF-to-motif mapping and see how many unique motifs are mapped to by our TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "save(TFs.to.motifs, motifs.to.TFs, file=\"Rdata_files/Tfmotifmap.Rdata\")\n",
    "# load(\"Rdata_files/Tfmotifmap.Rdata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "195"
      ],
      "text/latex": [
       "195"
      ],
      "text/markdown": [
       "195"
      ],
      "text/plain": [
       "[1] 195"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allmots <- c()\n",
    "\n",
    "for (TFname in names(TFs.to.motifs)) {\n",
    "    allmots  <-  c(allmots, TFs.to.motifs[[TFname]])\n",
    "}\n",
    "length(unique(allmots))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# write.table(unique(allmots), file=\"text_data_files/unique_motifs_for_CS_TFs.txt\", row.names=FALSE, col.names=FALSE, quote=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So as we can summarize our \"gold-standard\" ChipSeq data now as follows:\n",
    "\n",
    "* Our ChipSeq data contain 77 TFs, 60 of which we have a mapping for and 55 of which are JASPAR\n",
    "* Mapping our 55, we can get 195 unique motifs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## function to output df containing pos and neg examples for each TF\n",
    "\n",
    "Here's what we actually do for a given TF\n",
    "\n",
    "1. We pull out all ChipSeq hits for the TF and their associated regions\n",
    "2. Using our mapping, we look for ALL the occurences of the motifs for that TF in the FIMO database\n",
    "3. We find the overlaps of (1) and (2), so now we have the regions in FIMO where there's overlap with ChipSeq hits. And this is for a given TF\n",
    "4. Using our overlaps, we grab the FIMO motifs that are part of the overlaps; these are our \"positives\"\n",
    "5. Using a ratio of neg/pos, we find the \"positives\" for each motif of the TF and multiply it by the ratio. So if we have neg/pos = 0.5 and there are 10 hits (positives) of a particular motif, we want 5 negatives. \n",
    "6. We then find the actual \"negatives\" for each motif by finding those that don't pop up in the positives. \n",
    "7. Using our \"wanted\" number of negatives and our \"actual\", we take the smaller of these 2 (so we get our wanted number, unless there's fewer than that) and then sample the negatives to grab this many for each motif. \n",
    "8. Finally, we put the positive and negative together, with a column called \"cs_hit\" designating a positive or negative\n",
    "\n",
    "More succinctly, for each TF we:\n",
    "\n",
    "1. Find all the motifs for a TF from our mapping.\n",
    "2. Pull out all the places where these motifs in FIMO overlap the ChipSeq data\n",
    "3. Compile a dataset of places where we had hits and then a sampling of the same motifs in FIMO where we didn't overlap ChipSeq data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>loc</th><th scope=col>type</th><th scope=col>name</th><th scope=col>length</th><th scope=col>strand</th><th scope=col>sample_id</th><th scope=col>method</th><th scope=col>provenance</th><th scope=col>score1</th><th scope=col>score2</th><th scope=col>score3</th><th scope=col>score4</th><th scope=col>score5</th><th scope=col>score6</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1:1677841-1677991</td><td>chipseq.peak        </td><td>ATF2                </td><td>151                 </td><td>NA                  </td><td>pooled              </td><td>cusanovitch         </td><td>chipseq.minid.tbd   </td><td>155                 </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td></tr>\n",
       "\t<tr><td>chr1:1678101-1678251</td><td>chipseq.peak        </td><td>ATF2                </td><td>151                 </td><td>NA                  </td><td>pooled              </td><td>cusanovitch         </td><td>chipseq.minid.tbd   </td><td>131                 </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td></tr>\n",
       "\t<tr><td>chr1:1828321-1828471</td><td>chipseq.peak        </td><td>ATF2                </td><td>151                 </td><td>NA                  </td><td>pooled              </td><td>cusanovitch         </td><td>chipseq.minid.tbd   </td><td> 30                 </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td></tr>\n",
       "\t<tr><td>chr1:1828521-1828671</td><td>chipseq.peak        </td><td>ATF2                </td><td>151                 </td><td>NA                  </td><td>pooled              </td><td>cusanovitch         </td><td>chipseq.minid.tbd   </td><td> 53                 </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td></tr>\n",
       "\t<tr><td>chr1:1828761-1828911</td><td>chipseq.peak        </td><td>ATF2                </td><td>151                 </td><td>NA                  </td><td>pooled              </td><td>cusanovitch         </td><td>chipseq.minid.tbd   </td><td> 30                 </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td></tr>\n",
       "\t<tr><td>chr1:2255801-2255951</td><td>chipseq.peak        </td><td>ATF2                </td><td>151                 </td><td>NA                  </td><td>pooled              </td><td>cusanovitch         </td><td>chipseq.minid.tbd   </td><td>243                 </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td><td>NA                  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllll}\n",
       " loc & type & name & length & strand & sample\\_id & method & provenance & score1 & score2 & score3 & score4 & score5 & score6\\\\\n",
       "\\hline\n",
       "\t chr1:1677841-1677991 & chipseq.peak         & ATF2                 & 151                  & NA                   & pooled               & cusanovitch          & chipseq.minid.tbd    & 155                  & NA                   & NA                   & NA                   & NA                   & NA                  \\\\\n",
       "\t chr1:1678101-1678251 & chipseq.peak         & ATF2                 & 151                  & NA                   & pooled               & cusanovitch          & chipseq.minid.tbd    & 131                  & NA                   & NA                   & NA                   & NA                   & NA                  \\\\\n",
       "\t chr1:1828321-1828471 & chipseq.peak         & ATF2                 & 151                  & NA                   & pooled               & cusanovitch          & chipseq.minid.tbd    &  30                  & NA                   & NA                   & NA                   & NA                   & NA                  \\\\\n",
       "\t chr1:1828521-1828671 & chipseq.peak         & ATF2                 & 151                  & NA                   & pooled               & cusanovitch          & chipseq.minid.tbd    &  53                  & NA                   & NA                   & NA                   & NA                   & NA                  \\\\\n",
       "\t chr1:1828761-1828911 & chipseq.peak         & ATF2                 & 151                  & NA                   & pooled               & cusanovitch          & chipseq.minid.tbd    &  30                  & NA                   & NA                   & NA                   & NA                   & NA                  \\\\\n",
       "\t chr1:2255801-2255951 & chipseq.peak         & ATF2                 & 151                  & NA                   & pooled               & cusanovitch          & chipseq.minid.tbd    & 243                  & NA                   & NA                   & NA                   & NA                   & NA                  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "loc | type | name | length | strand | sample_id | method | provenance | score1 | score2 | score3 | score4 | score5 | score6 | \n",
       "|---|---|---|---|---|---|\n",
       "| chr1:1677841-1677991 | chipseq.peak         | ATF2                 | 151                  | NA                   | pooled               | cusanovitch          | chipseq.minid.tbd    | 155                  | NA                   | NA                   | NA                   | NA                   | NA                   | \n",
       "| chr1:1678101-1678251 | chipseq.peak         | ATF2                 | 151                  | NA                   | pooled               | cusanovitch          | chipseq.minid.tbd    | 131                  | NA                   | NA                   | NA                   | NA                   | NA                   | \n",
       "| chr1:1828321-1828471 | chipseq.peak         | ATF2                 | 151                  | NA                   | pooled               | cusanovitch          | chipseq.minid.tbd    |  30                  | NA                   | NA                   | NA                   | NA                   | NA                   | \n",
       "| chr1:1828521-1828671 | chipseq.peak         | ATF2                 | 151                  | NA                   | pooled               | cusanovitch          | chipseq.minid.tbd    |  53                  | NA                   | NA                   | NA                   | NA                   | NA                   | \n",
       "| chr1:1828761-1828911 | chipseq.peak         | ATF2                 | 151                  | NA                   | pooled               | cusanovitch          | chipseq.minid.tbd    |  30                  | NA                   | NA                   | NA                   | NA                   | NA                   | \n",
       "| chr1:2255801-2255951 | chipseq.peak         | ATF2                 | 151                  | NA                   | pooled               | cusanovitch          | chipseq.minid.tbd    | 243                  | NA                   | NA                   | NA                   | NA                   | NA                   | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  loc                  type         name length strand sample_id method     \n",
       "1 chr1:1677841-1677991 chipseq.peak ATF2 151    NA     pooled    cusanovitch\n",
       "2 chr1:1678101-1678251 chipseq.peak ATF2 151    NA     pooled    cusanovitch\n",
       "3 chr1:1828321-1828471 chipseq.peak ATF2 151    NA     pooled    cusanovitch\n",
       "4 chr1:1828521-1828671 chipseq.peak ATF2 151    NA     pooled    cusanovitch\n",
       "5 chr1:1828761-1828911 chipseq.peak ATF2 151    NA     pooled    cusanovitch\n",
       "6 chr1:2255801-2255951 chipseq.peak ATF2 151    NA     pooled    cusanovitch\n",
       "  provenance        score1 score2 score3 score4 score5 score6\n",
       "1 chipseq.minid.tbd 155    NA     NA     NA     NA     NA    \n",
       "2 chipseq.minid.tbd 131    NA     NA     NA     NA     NA    \n",
       "3 chipseq.minid.tbd  30    NA     NA     NA     NA     NA    \n",
       "4 chipseq.minid.tbd  53    NA     NA     NA     NA     NA    \n",
       "5 chipseq.minid.tbd  30    NA     NA     NA     NA     NA    \n",
       "6 chipseq.minid.tbd 243    NA     NA     NA     NA     NA    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "create.TF.df <- function(TF, neg.pos.ratio=10, verbose=FALSE) {\n",
    "    \n",
    "    # Note: this is how you make a DB connection in dplyr\n",
    "    db.fimo.dplyr <- src_postgres(drv=dbDriver(\"PostgreSQL\"),\n",
    "                                  user=\"trena\",\n",
    "                                  password=\"trena\",\n",
    "                                  dbname=\"fimo\",\n",
    "                                  host=\"whovian\",\n",
    "                                  port=\"5432\")\n",
    "    tbl.fimo.dplyr <- tbl(db.fimo.dplyr, \"fimo_hg38\")\n",
    "    \n",
    "    # regions locs we can compute on but hits have TF info so need both\n",
    "    # Here, we grab all hits for a TF, then grab the regions of those hits\n",
    "    chipseq.hits.TF <- subset(chipseq.hits, name == TF)\n",
    "    locs.TF <- chipseq.hits.TF$loc\n",
    "    chipseq.regions.TF <- subset(chipseq.regions, loc %in% locs.TF)\n",
    "    \n",
    "    # next step is slow, gives you some context\n",
    "    if (verbose == TRUE) {\n",
    "        if (length(TFs.to.motifs[[TF]])==1) {\n",
    "            message(paste(TF, \"- querying fimo database for\", length(TFs.to.motifs[[TF]]), \"motif\"))\n",
    "        } else {\n",
    "            message(paste(TF, \"- querying fimo database for\", length(TFs.to.motifs[[TF]]), \"motifs\"))\n",
    "        }       \n",
    "    }\n",
    "        \n",
    "    # this is the slow step -- doing SQL queries on tbl.fimo.dplyr = call to whole fimo database    \n",
    "    # need branch since %in% conversion to SQL doesn't work on length == 1\n",
    "    ## Basically: we find all instances of the TF's motif(s) in FIMO\n",
    "    if (length(TFs.to.motifs[[TF]]) > 1 ) {\n",
    "        fimo.motifs.for.TF <- as.tbl(as.data.frame(filter(tbl.fimo.dplyr, motifname %in% TFs.to.motifs[[TF]])))\n",
    "    } else {\n",
    "        fimo.motifs.for.TF <- as.tbl(as.data.frame(filter(tbl.fimo.dplyr, motifname  ==  TFs.to.motifs[[TF]])))\n",
    "    }\n",
    "    \n",
    "    # find intersect using fast genomic ranges data structure\n",
    "    # We make GR objects for the fimo motifs we just found and for the chipseq regions, then find their overlaps\n",
    "    gr.fimo.TF <- with(fimo.motifs.for.TF, GRanges(chrom, IRanges(start=start, end=endpos)))\n",
    "    gr.chipseq.TF <- with(chipseq.regions.TF, GRanges(chrom, IRanges(start=start, end=endpos)))\n",
    "    overlaps.gr.TF <- findOverlaps(gr.chipseq.TF, gr.fimo.TF, type=\"any\")\n",
    "    overlaps.TF <- as.tbl(as.data.frame(overlaps.gr.TF))\n",
    "    \n",
    "    # row numbers in fimo.motifs.for.TF where motifs overlap with chipseq peaks\n",
    "    positive.fimo.examples.rows.TF <- unique(overlaps.TF$subjectHits)\n",
    "    positive.fimo.examples.TF.df <- fimo.motifs.for.TF[positive.fimo.examples.rows.TF,]\n",
    "    \n",
    "    # figure out how many negative samples for each motif we want\n",
    "    tot.motif.counts.TF <- table(fimo.motifs.for.TF$motifname)\n",
    "    pos.motif.counts.TF <- table(fimo.motifs.for.TF[positive.fimo.examples.rows.TF,]$motifname)\n",
    "    nx.pos.motif.counts.TF <- pos.motif.counts.TF*neg.pos.ratio\n",
    "\n",
    "    # want neg samples in fimo.motifs.for.TF to be non overlapping with chipseq peaks or pos examples\n",
    "    neg.cands.for.single.TF.df <- subset(fimo.motifs.for.TF,\n",
    "                                         !(start %in% positive.fimo.examples.TF.df$start) &\n",
    "                                         !(endpos %in% positive.fimo.examples.TF.df$endpos))\n",
    "    neg.motif.counts.TF <- table(neg.cands.for.single.TF.df$motifname)\n",
    "\n",
    "    # don't try to sample more than the population\n",
    "    neg.sample.counts.TF <- pmin(nx.pos.motif.counts.TF,neg.motif.counts.TF)\n",
    "    \n",
    "    # for each motif this TF matches, sample some examples where no CS hit\n",
    "    negative.fimo.examples.TF.df <- tibble()\n",
    "    for (motname in names(neg.sample.counts.TF)) {\n",
    "        neg.cands.for.single.TF.df.single.motif <- subset(neg.cands.for.single.TF.df, motifname == motname)\n",
    "        negative.fimo.examples.TF.df <- rbind(negative.fimo.examples.TF.df,\n",
    "                                    sample_n(neg.cands.for.single.TF.df.single.motif, neg.sample.counts.TF[[motname]]))\n",
    "    }\n",
    "    \n",
    "    # annotate and collect all samples\n",
    "    positive.fimo.examples.TF.df <- as.tbl(cbind(positive.fimo.examples.TF.df, \"cs_hit\"=1))\n",
    "    negative.fimo.examples.TF.df <- as.tbl(cbind(negative.fimo.examples.TF.df, \"cs_hit\"=0))\n",
    "    all.fimo.examples.TF.df <- as.tbl(rbind(positive.fimo.examples.TF.df,negative.fimo.examples.TF.df))\n",
    "    \n",
    "    return(all.fimo.examples.TF.df)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create df of pos/neg samples for all TFs all together"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With our function defined, we'll now pull out the positive and negative samples for each TF; we're doing so with a neg/pos ration of 9. We're also doing this using a parallel strategy. The output gets saved in an `RData` file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# ratio of negative to positve examples in data set\n",
    "pnr=9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "\t<li><ol class=list-inline>\n",
       "\t<li>'GenomicRanges'</li>\n",
       "\t<li>'GenomeInfoDb'</li>\n",
       "\t<li>'IRanges'</li>\n",
       "\t<li>'S4Vectors'</li>\n",
       "\t<li>'BiocGenerics'</li>\n",
       "\t<li>'parallel'</li>\n",
       "\t<li>'stats4'</li>\n",
       "\t<li>'dplyr'</li>\n",
       "\t<li>'RPostgreSQL'</li>\n",
       "\t<li>'DBI'</li>\n",
       "\t<li>'stats'</li>\n",
       "\t<li>'graphics'</li>\n",
       "\t<li>'grDevices'</li>\n",
       "\t<li>'utils'</li>\n",
       "\t<li>'datasets'</li>\n",
       "\t<li>'methods'</li>\n",
       "\t<li>'base'</li>\n",
       "</ol>\n",
       "</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'parallel'\n",
       "\\item 'stats4'\n",
       "\\item 'dplyr'\n",
       "\\item 'RPostgreSQL'\n",
       "\\item 'DBI'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "2. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "3. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "4. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "5. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "6. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "7. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "8. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "9. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "10. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "11. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'parallel'\n",
       "7. 'stats4'\n",
       "8. 'dplyr'\n",
       "9. 'RPostgreSQL'\n",
       "10. 'DBI'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[2]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[3]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[4]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[5]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[6]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[7]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[8]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[9]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[10]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[11]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"parallel\"      \"stats4\"        \"dplyr\"        \n",
       " [9] \"RPostgreSQL\"   \"DBI\"           \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“closing unused connection 14 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 13 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 12 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 11 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 10 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 9 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 8 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 7 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 6 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 5 (<-localhost:11087)”Warning message:\n",
      "“closing unused connection 4 (<-localhost:11087)”"
     ]
    }
   ],
   "source": [
    "# make a cluster for parallel computing\n",
    "cl <- makePSOCKcluster(11)\n",
    "clusterEvalQ(cl, {\n",
    "    library(DBI)\n",
    "    library(RPostgreSQL)\n",
    "    library(dplyr)\n",
    "    library(GenomicRanges)\n",
    "})\n",
    "registerDoParallel(cl)\n",
    "\n",
    "# parallel loop over all TFs\n",
    "sorted.TF.names <- sort(names(TFs.to.motifs))\n",
    "N.TF <- length(sorted.TF.names)\n",
    "\n",
    "all.TF.df <- foreach(i.TF=1:N.TF,\n",
    "                     .inorder=FALSE,\n",
    "                     .combine=rbind,\n",
    "                     .packages=c(\"DBI\", \"RPostgreSQL\", \"dplyr\", \"GenomicRanges\")) %dopar% {    \n",
    "    TFname <- sorted.TF.names[[i.TF]]\n",
    "    create.TF.df(TFname, neg.pos.ratio=pnr, verbose=TRUE)\n",
    "}\n",
    "\n",
    "# clean up after parallel setup\n",
    "stopCluster(cl)\n",
    "registerDoSEQ()\n",
    "\n",
    "# save data\n",
    "fname=paste(\"/ssd/mrichard/data/all.TF.fimo.samples.ratio.\",pnr,\".df.RData\", sep=\"\")\n",
    "save(all.TF.df, file=fname)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## serial version of production loop\n",
    "\n",
    "This is the same thing, but done serially instead of with parallel computing. Not much to see here. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing TF 1 / 55\n",
      "ATF2 - querying fimo database for 8 motifs\n",
      "Processing TF 2 / 55\n",
      "ATF3 - querying fimo database for 3 motifs\n",
      "Processing TF 3 / 55\n",
      "BATF - querying fimo database for 2 motifs\n",
      "Processing TF 4 / 55\n",
      "BHLHE40 - querying fimo database for 1 motif\n",
      "Processing TF 5 / 55\n",
      "CEBPB - querying fimo database for 1 motif\n",
      "Processing TF 6 / 55\n",
      "CTCF - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 7 / 55\n",
      "E2F4 - querying fimo database for 7 motifs\n",
      "Processing TF 8 / 55\n",
      "EBF1 - querying fimo database for 1 motif\n",
      "Processing TF 9 / 55\n",
      "EGR1 - querying fimo database for 16 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 10 / 55\n",
      "ELF1 - querying fimo database for 21 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 11 / 55\n",
      "ELK1 - querying fimo database for 21 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 12 / 55\n",
      "ETS1 - querying fimo database for 1 motif\n",
      "Processing TF 13 / 55\n",
      "FOS - querying fimo database for 2 motifs\n",
      "Processing TF 14 / 55\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "FOXM1 - querying fimo database for 24 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 15 / 55\n",
      "IRF3 - querying fimo database for 5 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 16 / 55\n",
      "IRF4 - querying fimo database for 5 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 17 / 55\n",
      "JUND - querying fimo database for 8 motifs\n",
      "Processing TF 18 / 55\n",
      "MAX - querying fimo database for 11 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 19 / 55\n",
      "MEF2A - querying fimo database for 4 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 20 / 55\n",
      "MEF2C - querying fimo database for 4 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 21 / 55\n",
      "MTA3 - querying fimo database for 6 motifs\n",
      "Processing TF 22 / 55\n",
      "MXI1 - querying fimo database for 11 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 23 / 55\n",
      "MYC - querying fimo database for 2 motifs\n",
      "Processing TF 24 / 55\n",
      "NFATC1 - querying fimo database for 4 motifs\n",
      "Processing TF 25 / 55\n",
      "NFE2 - querying fimo database for 2 motifs\n",
      "Processing TF 26 / 55\n",
      "NFIC - querying fimo database for 4 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 27 / 55\n",
      "NFKB1 - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 28 / 55\n",
      "NFYA - querying fimo database for 2 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 29 / 55\n",
      "NFYB - querying fimo database for 2 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 30 / 55\n",
      "NR2C2 - querying fimo database for 2 motifs\n",
      "Processing TF 31 / 55\n",
      "NRF1 - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 32 / 55\n",
      "PAX5 - querying fimo database for 4 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 33 / 55\n",
      "PBX3 - querying fimo database for 8 motifs\n",
      "Processing TF 34 / 55\n",
      "POU2F2 - querying fimo database for 15 motifs\n",
      "Processing TF 35 / 55\n",
      "REST - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 36 / 55\n",
      "RFX5 - querying fimo database for 5 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 37 / 55\n",
      "RUNX3 - querying fimo database for 3 motifs\n",
      "Processing TF 38 / 55\n",
      "RXRA - querying fimo database for 6 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 39 / 55\n",
      "SIX5 - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 40 / 55\n",
      "SP1 - querying fimo database for 16 motifs\n",
      "Processing TF 41 / 55\n",
      "SPI1 - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 42 / 55\n",
      "SRF - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 43 / 55\n",
      "STAT1 - querying fimo database for 8 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 44 / 55\n",
      "STAT3 - querying fimo database for 8 motifs\n",
      "Processing TF 45 / 55\n",
      "STAT5A - querying fimo database for 8 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 46 / 55\n",
      "TAF1 - querying fimo database for 3 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 47 / 55\n",
      "TBP - querying fimo database for 1 motif\n",
      "Processing TF 48 / 55\n",
      "TCF12 - querying fimo database for 5 motifs\n",
      "Processing TF 49 / 55\n",
      "TCF3 - querying fimo database for 5 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 50 / 55\n",
      "USF1 - querying fimo database for 11 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 51 / 55\n",
      "USF2 - querying fimo database for 11 motifs\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 52 / 55\n",
      "YY1 - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 53 / 55\n",
      "ZBTB33 - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 54 / 55\n",
      "ZEB1 - querying fimo database for 1 motif\n",
      "Auto-disconnecting PostgreSQLConnection\n",
      "Processing TF 55 / 55\n",
      "ZNF143 - querying fimo database for 1 motif\n"
     ]
    }
   ],
   "source": [
    "# serial version -- hung on last TF, don't know why\n",
    "\n",
    "all.TF.df <- tibble()\n",
    "\n",
    "sorted.TF.names <- sort(names(TFs.to.motifs))\n",
    "for (TFname in sorted.TF.names) {\n",
    "    \n",
    "    TFnum <- which(sorted.TF.names %in% TFname)\n",
    "    message(paste(\"Processing TF\", TFnum,\"/\", length(sorted.TF.names)))\n",
    "    TF.df <- create.TF.df(TFname, neg.pos.ratio=9, verbose=TRUE)\n",
    "    all.TF.df <- rbind(all.TF.df,TF.df)\n",
    "    \n",
    "    save(all.TF.df, file=\"/ssd/mrichard/data/all.TF.fimo.samples.ratio.9.df.RData\")\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
